faunadb-java-dsl/src/main/java/com/faunadb/client/types/Encoder.java

```diff
@@ -0,0 +1,130 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Stack;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Lists.newArrayListWithCapacity;
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+/**
+ * FaunaDB object to {@link Value} encoder.
+ */
+public final class Encoder {
+  private Encoder() {
+  }
+
+  private final static Stack<Object> stack = new Stack<>();
+
+  /**
+   * Encode the specified object into a corresponding FaunaDB value.
+   * <p>
+   * <pre>{@code
+   * encode(10) => new LongV(10)
+   * encode(3.14) => new DoubleV(3.14)
+   * encode(true) => BooleanV.TRUE
+   * encode(null) => NullV.NULL
+   * encode("a string") => new StringV("a string")
+   * encode(new int[] {1, 2}) => new ArrayV(Arrays.asList(1, 2))
+   * encode(Arrays.asList(1, 2)) => new ArrayV(Arrays.asList(1, 2))
+   * encode(new byte[] {1, 2}) => new ByteV(new byte[] {1, 2})
+   * encode(user) => new ObjectV(ImmutableMap.of("user_name", new StringV("john"), "password", new StringV("s3cr3t")))
+   * }</pre>
+   *
+   * @param obj Any instance of user defined classes, primitive values or any
+   *            generic collection like {@link java.util.List}, {@link java.util.Set} or {@link java.util.Map}
+   * @return A FaunaDB {@link Value} corresponding to the given argument
+   */
+  public static Result<Value> encode(Object obj) {
+    if (obj == null)
+      return Result.<Value>success(Value.NullV.NULL);
+
+    if (Value.class.isInstance(obj))
+      return Result.success((Value) obj);
+
+    if (contains(obj))
+      return Result.fail(format("Self reference loop detected for object `%s`", obj));
+
+    try {
+      stack.push(obj);
+
+      return encodeIntern(obj);
+    } finally {
+      stack.pop();
+    }
+  }
+
+  private static boolean contains(Object obj) {
+    int size = stack.size();
+
+    for (int i = 0; i < size; i++) {
+      if (stack.get(i) == obj)
+        return true;
+    }
+
+    return false;
+  }
+
+  private static Result<Value> encodeIntern(Object obj) {
+    Class<?> clazz = obj.getClass();
+
+    if (clazz.isEnum())
+      return Result.<Value>success(Enums.getEncodingMap((Class<Enum>) clazz).get(obj));
+
+    if (clazz == byte[].class)
+      return Codec.BYTES.encode((byte[]) obj);
+
+    if (clazz.isArray())
+      return wrapArray(obj);
+
+    if (Iterable.class.isAssignableFrom(clazz))
+      return wrapIterable((Iterable<?>) obj);
+
+    if (Map.class.isAssignableFrom(clazz))
+      return wrapMap((Map<?, ?>) obj);
+
+    return ((Codec<Object>) Codecs.getCodec(clazz)).encode(obj);
+  }
+
+  private static Result<Value> wrapMap(Map<?, ?> obj) {
+    Map<String, Value> values = newHashMap();
+
+    for (Entry<?, ?> entry : obj.entrySet()) {
+      values.put(entry.getKey().toString(), encode(entry.getValue()).get());
```

Comment:
I think we must enforce `String` keys here because we can't materialize the same object back. For example: `Map<User, Token>` will be serialized as `Map<String, Map<String, Value>>` but we can't deserialize a `String` to a `User` when the data comes back.